### DISCUSSION TITLE
Rethinking `fly launch`

### DISCUSSION THREAD

**Discussion by rubys** [remix-run/remix#6191] [Proposals]
My name is Sam Ruby and I work for fly.io.  I'm planning on going to [RemixConf 2023](https://remix.run/conf/2023), and hope to start a discussion on:

  * Replacing [these manual deploy steps](https://github.com/remix-run/indie-stack/blob/main/README.md?plain=1#L86-L129) with `fly launch`.
 * Having Fly and Remix converge on a single [dockerfile generator](https://github.com/fly-apps/dockerfile-node)
 * Prepping for [LiteFS](https://github.com/superfly/litefs#readme)

Background: I have a similar [generator for Rails](https://github.com/rubys/dockerfile-rails) and contributed heavily to their [docker template](https://github.com/rails/rails/blame/main/railties/lib/rails/generators/rails/app/templates/Dockerfile.tt).  For Rails, the generator is called during `fly launch`.  I can do the same for Remix.  The process can do anything a script can do, for example [setting of session secrets](https://github.com/superfly/flyctl/blob/0f34df0096fed901885f911896f786fdb58a4093/scanner/node.go#L110-L123).

**Comment by rubys** [comment]
Related discussions: [fly community](https://community.fly.io/t/error-failed-to-spawn-command-start-sh-permission-denied-os-error-13/12621/11?u=rubys) and [remix github](https://github.com/remix-run/indie-stack/issues/231#issuecomment-1531322068).

**Reply by MichaelDeBoey** [reply]
These are probably related as well & could probably be automatically fixed by the generator detecting certain situations:
- https://github.com/remix-run/indie-stack/issues/200
- https://github.com/remix-run/indie-stack/issues/213
- https://github.com/remix-run/blues-stack/issues/138
- https://github.com/remix-run/blues-stack/issues/143
- https://github.com/remix-run/blues-stack/issues/170

**Reply by rubys** [reply]
Over the next few days, I plan to make [dockerfile-node](https://github.com/fly-apps/dockerfile-node) the dockerfile generator for Node applications on fly.io.  I will set things up so that it won't replace any Dockerfile you might have, I will be creating web pages showing how to use it to replace your Dockerfile to solve common problems (examples: [running on windows](https://community.fly.io/t/error-failed-to-spawn-command-start-sh-permission-denied-os-error-13/12621), handle [legacy peer dependencies](https://community.fly.io/t/error-while-deploying-node-js-app/12575), and [running out of memory](https://github.com/remix-run/indie-stack/issues/200)).

It is my hope to meet with people at RemixConf and use that time to make my case on replacing the Dockerfile templates that exist in Remix with this generator.

As to [no region defined](https://github.com/remix-run/indie-stack/issues/213) and [fly apps create](https://github.com/remix-run/blues-stack/issues/138), fly.io recently made a rather significant runtime change.  Some issues with the `fly.toml` created by remix:
 * [experimental] section with a cmd =
 * No primary_region
  * [[services]] section rather than [http_service] with no auto_stop_machines or auto_start_machines 

My recommendation will be that Remix stop creating a fly.toml, and update the deployment instructions to use `fly launch` instead.  I'd like to talk through this and will either make the changes needed to `fly launch` to support this, or champion the need for changes within fly.io.

Finally, on the [seed issue](https://github.com/remix-run/blues-stack/issues/143), if you want to be able to seed your production databases, you will need to make sure that all of the dependencies necessary to do so are present in the production environment.

**Reply by MichaelDeBoey** [reply]
Some extra related links:
- https://community.fly.io/t/my-postgres-database-is-down-for-no-reason/12664
- https://github.com/remix-run/indie-stack/pull/232
- https://github.com/remix-run/indie-stack/pull/234
- https://github.com/remix-run/blues-stack/issues/88
- https://github.com/remix-run/blues-stack/pull/101
- https://github.com/remix-run/blues-stack/issues/182
- https://github.com/remix-run/blues-stack/pull/183
- https://github.com/remix-run/blues-stack/pull/184

**Reply by rubys** [reply]
> https://community.fly.io/t/my-postgres-database-is-down-for-no-reason/12664

[this line](https://github.com/remix-run/blues-stack/blob/main/app/db.server.ts#L37) is likely the culprit, this is something flycast should handle.

> https://github.com/remix-run/indie-stack/pull/232
> https://github.com/remix-run/blues-stack/pull/183

That's what `fly scale` and the `--swap` flag are for.  If it is an app you care about, you want real memory.  If it is a hobby app, virtual memory will do.

> https://github.com/remix-run/blues-stack/pull/101
> https://github.com/remix-run/blues-stack/issues/88

This seems related to the first issue above.  I believe that `process.env.PRIMARY_REGION` is an apps V2 feature, so if the user has an older fly organization that hasn't been migrated, you will see `undefined`.

> https://github.com/remix-run/blues-stack/pull/184

This does several changes.

 * `install` => `ci` ; this is a very good change, and dockerfile-node should be fixed to use this
 * `--production=false` => `--include=dev`; this also is a good change, and dockerfile-node already does this
 * `--ignore-scripts` -- I'm curious as to why that is needed.  If it is not commonly needed, then I would be inclined to see it added as an option to dockerfile-node.

---

Meta comment: fly.io is still changing rapidly - and that's a good thing!  But it keeps me busy - I've been focusing on Rails this past year or so and now am starting to expand to include Node frameworks.  I have a high bandwidth communication with the product team and can react quickly to changes that are made to the platform.  Each framework having fly specific code -- like remix currently does in its product -- leads to problems like these.  It would be much better if that were factored out into [dockerfile-node](https://github.com/fly-apps/dockerfile-node) and `fly launch` which can be kept up to date with platform changes.

That's why I'm flying out to remix-conf this week and hope to have a high bandwidth conversation on the matter with @kentcdodds and anybody else interested in making this happen.

**Reply by MichaelDeBoey** [reply]
@rubys I appreciate all the work you're doing on the new `dockerfile-node` library 🙏
Just wanted to mention all the Fly-related issues/PRs I come across as they can potentially improve the lib (if things aren't implemented yet)

> It would be much better if that were factored out into [dockerfile-node](https://github.com/fly-apps/dockerfile-node) and `fly launch` which can be kept up to date with platform changes.

It would indeed be better if we could just remove `Dockerfile`, `fly.toml`, ... from our repos and just call `fly launch` (maybe with some flags) without having to handle all the other stuff ourselves 👀

**Comment by rubys** [comment]
Adding, for completeness: https://github.com/remix-run/blues-stack/pull/187

Also, the issues here also affect the [newly announced epic stack](https://github.com/epicweb-dev/epic-stack/discussions/22).

**Comment by abetoots** [comment]
hi @rubys just wanted to check if there are any updates regarding the conversation with remix and @kentcdodds ?

**Comment by rubys** [comment]
@abetoots Not really. Epic stack is still not using fly launch for [deployment](https://github.com/epicweb-dev/epic-stack/blob/main/docs/deployment.md).  We've made a lot of improvements, and more are in the queue.  I'd love for Remix to be able to take advantage of those improvements.

cc: @anniebabannie

**Comment by brophdawg11** [comment]
This discussion is being closed as part of the Remix v3 repository cleanup.

Remix v1/v2 was [upstreamed into React Router v7](https://remix.run/blog/wake-up-remix), so if this proposal is still relevant, please reopen or continue the discussion in the [React Router](https://github.com/remix-run/react-router) repository.
